dcap/types/src/quote3.rs

```diff
@@ -4,13 +4,48 @@
 
 #[cfg(feature = "alloc")]
 use alloc::vec::Vec;
+use mc_sgx_core_types::{QuoteNonce, ReportData};
+use sha2::{Digest, Sha256};
+use subtle::ConstantTimeEq;
 
 /// Quote version 3
 #[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
 pub struct Quote3<T> {
     bytes: T,
 }
 
+impl<T: AsRef<[u8]>> Quote3<T> {
+    /// Verify the provided `nonce` matches the one in `report_data`
+    ///
+    /// When a nonce is passed to the quote generation, a QE report will be
+    /// returned where the report data is `SHA256(nonce||quote)||32-0x00's`.
+    /// This will verify that the `nonce` and this quote instance match the
+    /// provided `report_data`
+    ///
+    /// > Note: This report data is *not* the QE report data in the quote, it is
+    ///     part of the report info returned from SGX SDK quote generation.
+    ///
+    /// # Arguments
+    /// * `nonce` - The nonce believed to be in the `report_data`
+    /// * `report_data` - The report data to verify matches the `nonce` and this
+    ///     quote instance.
+    ///
+    /// Returns `true` if the `report_data` matches the `nonce` and this quote
+    /// instance.  Returns `false` if they differ.
+    ///
+    pub fn verify_nonce(&self, nonce: &QuoteNonce, report_data: &ReportData) -> bool {
+        let mut hasher = Sha256::new();
+        hasher.update(nonce);
+        hasher.update(&self.bytes);
+        let hash = hasher.finalize();
+
+        let mut data = [0u8; ReportData::SIZE];
```

Comment:
The docs on [`sgx_ql_get_quote()`](https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/master/QuoteGeneration/quote_wrapper/ql/sgx_dcap_ql_wrapper.cpp#L411) says  
> SHA256(*p_nonce||*p_quote)||32-0x00's.  

So implemented this to also compare the 32 0s